File: AbstractFactory.js
/*****
 * Abstract Factory
 * Create a series of related objects without specifying a specific class
 * to produce a series of objects.
*****/


class ProductA { method() {} }

class ProductA1 extends ProductA { method() { console.log("Product A1"); } }

class ProductA2 extends ProductA { method() { console.log("Product A2"); } }

class ProductB { method() {} }

class ProductB1 extends ProductB { method() { console.log("Product B1"); } }

class ProductB2 extends ProductB { method() { console.log("Product B2"); } }

class AbstractFactory {
    createA() {}
    createB() {}
}

class Factory1 extends AbstractFactory {
    createA() { return new ProductA1(); }
    createB() { return new ProductB1(); }
}

class Factory2 extends AbstractFactory {
    createA() { return new ProductA2(); }
    createB() { return new ProductB2(); }
}


const f1 = new Factory1();
const f2 = new Factory2();

const a1 = f1.createA();
const b1 = f1.createB();
const a2 = f2.createA();
const b2 = f2.createB();

a1.method();
b1.method();
a2.method();
b2.method();

File: AbstractFactory.cpp
/*****
 * Abstract Factory
 * Create a series of related objects without specifying a specific class
 * to produce a series of objects.
*****/

#include <iostream>
#include <memory>

using namespace std;


class ProductA { public: virtual void method() = 0; };

class ProductA1 : public ProductA { public: void method() override { cout << "Product A1" << endl; } };

class ProductA2 : public ProductA { public: void method() override { cout << "Product A2" << endl; } };

class ProductB { public: virtual void method() = 0; };

class ProductB1 : public ProductB { public: void method() override { cout << "Product B1" << endl; } };

class ProductB2 : public ProductB { public: void method() override { cout << "Product B2" << endl; } };

class AbstractFactory {
    public:
        virtual unique_ptr<ProductA> createA() = 0;
        virtual unique_ptr<ProductB> createB() = 0;
};

class Factory1 : public AbstractFactory {
    public:
        unique_ptr<ProductA> createA() override { return make_unique<ProductA1>(); }
        unique_ptr<ProductB> createB() override { return make_unique<ProductB1>(); }
};

class Factory2 : public AbstractFactory {
    public:
        unique_ptr<ProductA> createA() override { return make_unique<ProductA2>(); }
        unique_ptr<ProductB> createB() override { return make_unique<ProductB2>(); }
};


int main() {
    unique_ptr<AbstractFactory> f1 = make_unique<Factory1>();
    unique_ptr<AbstractFactory> f2 = make_unique<Factory2>();
    
    auto a1 = f1->createA();
    auto b1 = f1->createB();
    auto a2 = f2->createA();
    auto b2 = f2->createB();
    
    a1->method();
    b1->method();
    a2->method();
    b2->method();

    return 0;
}
